compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -5856,31 +5856,92 @@ private void handleSafeNavigation(BLangAccessExpression accessExpr, BType type,
 
         BLangMatch matchStmt = ASTBuilderUtil.createMatchStatement(accessExpr.pos, accessExpr.expr, new ArrayList<>());
 
+        boolean allRecords = false;
+        BUnionType errorLiftedType = BUnionType.create(null, new LinkedHashSet<>());
+        if (accessExpr.expr.type.tag == TypeTags.UNION) {
+            LinkedHashSet<BType> memTypes = new LinkedHashSet<>(((BUnionType) accessExpr.expr.type).getMemberTypes());
+            errorLiftedType = BUnionType.create(null, memTypes);
+            allRecords = checkForRecords(memTypes);
+        }
+
         // Add pattern to lift nil
         if (accessExpr.nilSafeNavigation) {
             matchStmt.patternClauses.add(getMatchNullPattern(accessExpr, tempResultVar));
             matchStmt.type = type;
+            errorLiftedType.remove(symTable.nilType);
         }
 
         // Add pattern to lift error, only if the safe navigation is used
         if (accessExpr.errorSafeNavigation) {
             matchStmt.patternClauses.add(getMatchErrorPattern(accessExpr, tempResultVar));
             matchStmt.type = type;
             matchStmt.pos = accessExpr.pos;
+            errorLiftedType.remove(symTable.errorType);
+        }
+
+        BLangMatchTypedBindingPatternClause successPattern = null;
+        Name field = getFieldName(accessExpr);
+        if (field == Names.EMPTY) {
+            successPattern = getSuccessPattern(accessExpr, tempResultVar,
+                    accessExpr.errorSafeNavigation);
+            matchStmt.patternClauses.add(successPattern);
+            pushToMatchStatementStack(matchStmt, accessExpr, successPattern);
+            return;
+        }
 
+        if (allRecords) {
+            for (BType memberType : errorLiftedType.getMemberTypes()) {
+                BSymbol fieldSymbol = symResolver.resolveStructField(accessExpr.pos, this.env, field,
+                        memberType.tsymbol);
```

Comment:
Do we need to do this? Can we not just iterate through the fields of the `BRecordType` and figure out if the field exists? 

`symResolver.resolveStructField` could log errors right, even thought it may not atm?

Reply:
@pubudu91 thoughts?

Reply:
This would log an error if at least one record type doesn't have the field name given by `field` right?

Reply:
And yeah, we don't usually use this to lookup whether the a record has a field by the specified name.

Reply:
Done